packages/prompts/src/index.ts

```diff
@@ -727,22 +727,28 @@ export const stream = {
 
 export interface SpinnerOptions {
 	indicator?: 'dots' | 'timer';
+	onCancel?: () => void;
 }
 
-export const spinner = ({ indicator = 'dots' }: SpinnerOptions = {}) => {
+export const spinner = ({ indicator = 'dots', onCancel }: SpinnerOptions = {}) => {
 	const frames = unicode ? ['◒', '◐', '◓', '◑'] : ['•', 'o', 'O', '0'];
 	const delay = unicode ? 80 : 120;
 	const isCI = process.env.CI === 'true';
 
 	let unblock: () => void;
 	let loop: NodeJS.Timeout;
 	let isSpinnerActive = false;
+	let isCancelled = false;
 	let _message = '';
 	let _prevMessage: string | undefined = undefined;
 	let _origin: number = performance.now();
 
 	const handleExit = (code: number) => {
 		const msg = code > 1 ? 'Something went wrong' : 'Canceled';
+		isCancelled = code === 1;
+		if (isCancelled && typeof onCancel === 'function') {
+			onCancel();
+		}
 		if (isSpinnerActive) stop(msg, code);
```

Comment:
since we will stop it here anyway, maybe we should call `onCancel` _after_ this line

that would mean the spinner has been cancelled, therefore stopped, and we've called `onCancel` to let you know

Reply:
hmm, you're right. what if we approach this like:

1. detecting cancellation
2. stopping the spinner 
3. call the onCancel callback

```ts
	const handleExit = (code: number) => {
		const msg = code > 1 ? 'Something went wrong' : 'Canceled';
		isCancelled = code === 1;
		if (isSpinnerActive) {
			stop(msg, code);
			if (isCancelled && typeof onCancel === 'function') {
				onCancel();
			}
		}
	};
```

this would ensure the spinner is properly stopped before the callback is invoked.

what do you think?

Reply:
that makes sense to me 👍 

you can't really cancel a spinner you haven't started, so i think you're right